src/validation.rs

```diff
@@ -0,0 +1,147 @@
+// We can discuss if that is better or not.
+
+pub struct TransactionValidator {}
+
+impl TransactionValidator {
+    // note: for spent_proofs to validate, the mint_verifier must have/know the spentbook section's public key.
+    pub fn validate<K: KeyManager>(
+        mint_verifier: &K,
+        transaction: &RingCtTransaction,
+        transaction_sigs: &BTreeMap<KeyImage, (PublicKey, Signature)>,
+        spent_proofs: &BTreeSet<SpentProof>,
+    ) -> Result<(), Error> {
+        // Do quick checks first to reduce potential DOS vectors.
+
+        if transaction_sigs.len() < transaction.mlsags.len() {
+            return Err(Error::MissingSignatureForInput);
+        }
+
+        let tx_hash = Hash::from(transaction.hash());
+
+        // Verify that each input has a corresponding valid mint signature.
+        for (key_image, (mint_key, mint_sig)) in transaction_sigs.iter() {
+            if !transaction
+                .mlsags
+                .iter()
+                .any(|m| m.key_image == *key_image.as_ref())
+            {
+                return Err(Error::UnknownInput);
+            }
+
+            mint_verifier
+                .verify(&tx_hash, mint_key, mint_sig)
+                .map_err(|e| Error::Signing(e.to_string()))?;
+        }
+
+        Self::validate_without_sigs_internal(mint_verifier, transaction, tx_hash, spent_proofs)
+    }
+
+    pub fn validate_without_sigs<K: KeyManager>(
+        mint_verifier: &K,
+        transaction: &RingCtTransaction,
+        spent_proofs: &BTreeSet<SpentProof>,
+    ) -> Result<(), Error> {
+        let tx_hash = Hash::from(transaction.hash());
+        Self::validate_without_sigs_internal(mint_verifier, transaction, tx_hash, spent_proofs)
+    }
+
+    fn validate_without_sigs_internal<K: KeyManager>(
+        mint_verifier: &K,
+        transaction: &RingCtTransaction,
+        transaction_hash: Hash,
+        spent_proofs: &BTreeSet<SpentProof>,
+    ) -> Result<(), Error> {
+        if transaction.mlsags.is_empty() {
+            return Err(Error::TransactionMustHaveAnInput);
+        } else if spent_proofs.len() != transaction.mlsags.len() {
+            return Err(Error::SpentProofInputMismatch);
+        }
+
+        // Verify that each KeyImage is unique in this tx.
+        let keyimage_unique: BTreeSet<KeyImage> = transaction
+            .mlsags
+            .iter()
+            .map(|m| m.key_image.into())
+            .collect();
+        if keyimage_unique.len() != transaction.mlsags.len() {
+            return Err(Error::KeyImageNotUniqueAcrossInputs);
+        }
+
+        // Verify that each pubkey is unique in this transaction.
+        let pubkey_unique: BTreeSet<KeyImage> = transaction
+            .outputs
+            .iter()
+            .map(|o| (*o.public_key()).into())
+            .collect();
+        if pubkey_unique.len() != transaction.outputs.len() {
+            return Err(Error::PublicKeyNotUniqueAcrossOutputs);
+        }
```

Comment:
This check should be moved to `RingCtTransaction::verify()` if it's not already there.

Reply:
created #130